lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -97,6 +97,144 @@ static Value createTMTensorScanOp(
+    RankedTensorType srcType = src.getType().cast<RankedTensorType>();
+    int64_t indexRank = indexType.getRank();
+    if (selfType.getRank() != indexRank || indexRank != srcType.getRank())
+      return rewriter.notifyMatchFailure(
+          op,
+          "self, index and src should all have the same number of dimensions.");
+
+    Value dimValue = op.dim();
+    int64_t dim;
+    if (!matchPattern(dimValue, m_TorchConstantInt(&dim)))
+      return op.emitError("unimplemented: dim is not constant");
+
+    // form inputs to the tm_tensor.scatter op
+    Value indexSize = getTensorSize(rewriter, loc, index);
+    indexSize = castIntToIndex(rewriter, loc, indexSize);
+    SmallVector<Value> indexShape = getTensorSizes(rewriter, loc, index);
+    Value cstOne = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+
+    // all the indices slices shapes will be (indexSize, 1)
+    SmallVector<Value> indSliceShape({indexSize, cstOne});
+    Value indSlice = createZeroInitTensor(rewriter, loc, indSliceShape,
+                                          rewriter.getI32Type());
+    SmallVector<Value> outputs(indexRank, indSlice);
+    // updates tensor will be of shape (indexSize).
+    outputs.push_back(createZeroInitTensor(rewriter, loc, {indexSize},
+                                           selfType.getElementType()));
+    SmallVector<Type> outputsType(indexRank, indSlice.getType());
+    outputsType.push_back(outputs[indexRank].getType());
+
+    // create mapping
+    SmallVector<AffineExpr> indSliceExpr;
+    indSliceExpr.push_back(rewriter.getAffineDimExpr(0));
+    indSliceExpr.push_back(rewriter.getAffineConstantExpr(0));
+    SmallVector<AffineMap> mapping(
+        indexRank, AffineMap::get(/*dimCount=*/1, /*symbolCount=*/0,
+                                  indSliceExpr, context));
+    // mapping for updates
+    mapping.push_back(rewriter.getDimIdentityMap());
+    SmallVector<StringRef> iteratorTypes({getParallelIteratorTypeName()});
+    auto tmTensorScatterInputs =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, outputsType, ValueRange(), outputs, mapping, iteratorTypes,
+                [&](OpBuilder &b, Location loc, ValueRange args) {
+                  SmallVector<Value> indices(indexRank);
+                  // 1D to n-D mapping
+                  Value ind = b.create<linalg::IndexOp>(loc, 0);
+                  for (int i = indexRank - 1; i >= 0; i--) {
+                    indices[i] =
+                        b.create<arith::RemSIOp>(loc, ind, indexShape[i]);
+                    ind = b.create<arith::DivSIOp>(loc, ind, indexShape[i]);
+                  }
+                  // extrace values
+                  Value extractIndexValue =
+                      b.create<tensor::ExtractOp>(loc, index, indices);
+                  Value extractSrcValue =
+                      b.create<tensor::ExtractOp>(loc, src, indices);
```

Comment:
I think this should also be casted to match the input tensor dtype. Can you also add a test where the input tensor is an `int` tensor and the `src` is a `float` tensor to see if the casting is being handled correctly?